### PR TITLE
add heter ps mode

### PR DESCRIPTION
**Pull request by Thunderbrook**
### PR types
New features

### PR changes
Others 

### Describe
add heter ps distribute mode in padde

### CODE REVIEW

**Review by fuyinno4 — DISMISSED** [review]

**Review by fuyinno4 — DISMISSED** [review]

File: python/paddle/fluid/incubate/fleet/base/role_maker.py [hunk 1]
```diff
@@ -988,6 +989,147 @@ def __start_kv_server(self, http_server_d, size_d):
         http_server.stop()
 
 
+class HeterRoleMaker(GeneralRoleMaker):
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/base/role_maker.py:992]: You may need to inherit from PaddleCloudRoleMaker

File: python/paddle/fluid/incubate/fleet/utils/fleet_util.py [hunk 2]
```diff
@@ -1615,3 +1616,123 @@ def parse_program_proto(self, prog_path, is_text, output_dir):
         """
         program = utils.load_program(prog_path, is_text)
         utils.parse_program(program, output_dir)
+
+    def split_program_by_device(self, program):
```
Comment (guru4elephant) [inline python/paddle/fluid/incubate/fleet/utils/fleet_util.py:1620]: API命名有点宽泛，另外是否作为隐藏API比较合理

**Review by guru4elephant — APPROVED** [review]
LGTM

File: python/paddle/fluid/incubate/fleet/base/role_maker.py (hunk 1 above)
Comment (Thunderbrook) [reply python/paddle/fluid/incubate/fleet/base/role_maker.py:992]: 之后会合并~

File: python/paddle/fluid/incubate/fleet/utils/fleet_util.py (hunk 2 above)
Comment (Thunderbrook) [reply python/paddle/fluid/incubate/fleet/utils/fleet_util.py:1620]: 会在下个pr修复~

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by saxon-zh — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**